src/interpreter/InterpreterRelation.h

```diff
@@ -31,244 +31,405 @@
+    void insert(const InterpreterRelation<Arity, Structure>& other) {
+        for (const auto& tuple : other.scan()) {
+            this->insert(tuple);
+        }
+    }
 
     /**
-     * Swaps the content of this and the given relation, including the
-     * installed indexes.
+     * Tests whether this relation contains the given tuple.
      */
-    void swap(InterpreterRelation& other);
+    bool contains(const Tuple& tuple) const {
+        return main->contains(tuple);
+    }
 
     /**
-     * Set level
+     * Tests whether this relation contains any element between the given boundaries.
      */
-    void setLevel(size_t level) {
-        this->level = level;
+    bool contains(const size_t& indexPos, const Tuple& low, const Tuple& high) const {
+        return indexes[indexPos]->contains(low, high);
     }
 
     /**
-     * Return the level of the relation.
+     * Obtains a pair of iterators to scan the entire relation.
+     *
+     * Return 'raw iterator' that returns tuples in undecoded form.
      */
-    size_t getLevel() const;
+    souffle::range<iterator> scan() const {
+        return main->scan();
+    }
 
     /**
-     * Return the relation name.
+     * Returns a partitioned list of iterators for parallel computation
      */
-    const std::string& getName() const;
+    std::vector<souffle::range<iterator>> partitionScan(size_t partitionCount) const {
+        return main->partitionScan(partitionCount);
+    }
 
     /**
-     * Return the attribute types
+     * Obtains a pair of iterators covering the interval between the two given entries.
      */
-    const std::vector<std::string>& getAttributeTypes() const;
+    souffle::range<iterator> range(const size_t& indexPos, const Tuple& low, const Tuple& high) const {
+        return indexes[indexPos]->range(low, high);
+    }
 
     /**
-     * Return arity
+     * Returns a partitioned list of iterators coving elements in range [low, high]
      */
-    size_t getArity() const {
-        return arity;
+    std::vector<souffle::range<iterator>> partitionRange(
+            const size_t& indexPos, const Tuple& low, const Tuple& high, size_t partitionCount) const {
+        return indexes[indexPos]->partitionRange(low, high, partitionCount);
     }
 
     /**
-     * Return arity
+     * Swaps the content of this and the given relation, including the
+     * installed indexes.
      */
-    size_t getAuxiliaryArity() const;
+    void swap(InterpreterRelation<Arity, Structure>& other) {
+        indexes.swap(other.indexes);
+    }
 
     /**
      * Return number of tuples in relation (full-order)
      */
-    size_t size() const;
+    size_t __size() const {
+        return main->size();
+    }
 
     /**
      * Check if the relation is empty
      */
-    bool empty() const;
+    bool empty() const {
+        return main->empty();
+    }
 
     /**
      * Clear all indexes
      */
-    virtual void purge();
+    void __purge() {
```

Comment:
Is there a good reason for using `__purge` instead of `purge` (and likewise for `__size` above)?

Reply:
`purge` is used in the `InterpreterProgInterface` and provenance program when the exact type of the target relation is not known ( or cannot be provided by the user), hence is a virtual function.
`__purge` is used inside the interpreter execution when the exact type of the target relation is known.